parsl/concurrent/__init__.py

```diff
@@ -1,35 +1,62 @@
 """Interfaces modeled after Python's `concurrent library <https://docs.python.org/3/library/concurrent.html>`_"""
 import time
 from concurrent.futures import Executor
-from typing import Callable, Dict, Iterable, Iterator, Optional
+from contextlib import AbstractContextManager
+from typing import Callable, Dict, Iterable, Iterator, Literal, Optional
 from warnings import warn
 
-from parsl import Config, DataFlowKernel
+from parsl import Config, DataFlowKernel, load
 from parsl.app.python import PythonApp
 
 
-class ParslPoolExecutor(Executor):
+class ParslPoolExecutor(Executor, AbstractContextManager):
     """An executor that uses a pool of workers managed by Parsl
 
     Works just like a :class:`~concurrent.futures.ProcessPoolExecutor` except that tasks
     are distributed across workers that can be on different machines.
-    Create a new executor by supplying a Parsl :class:`~parsl.Config` object to define
-    how to create new workers, Parsl will set up and tear down workers on your behalf.
+
+    Create a new executor using one of two methods:
+
+    1. Supplying a Parsl :class:`~parsl.Config` that defines how to create new workers.
+       The executor will start a new Parsl Data Flow Kernel (DFK) when it is entered as a context manager.
+
+    2. Supplying an already-started Parsl :class:`~parsl.DataFlowKernel` (DFK).
+       The executor assumes you will start and stop the Parsl DFK outside the Executor.
 
     Note: Parsl does not support canceling tasks. The :meth:`map` method does not cancel work
     when one member of the run fails or a timeout is reached
     and :meth:`shutdown` does not cancel work on completion.
     """
 
-    def __init__(self, config: Config):
+    def __init__(self, config: Config | None = None, dfk: DataFlowKernel | None = None, executors: Literal['all'] | list[str] = 'all'):
         """Create the executor
 
         Args:
             config: Configuration for the Parsl Data Flow Kernel (DFK)
+            dfk: DataFlowKernel of an already-started parsl
+            executors: List of executors to use for supplied functions
         """
+        if (config is not None) and (dfk is not None):
+            raise ValueError('Specify only one of config or dfk')
+        if (config is None) and (dfk is None):
+            raise ValueError('Must specify one of config or dfk')
```

Comment:
I was going to say "oh, for an xor operator". But then I remembered Python has one and this double test could turn into `assert one ^ the other, raise "you must specify exactly one of:"`